### ISSUE TITLE
Shouldn't ImageSurface has an AsMut<Surface> impl?

### ISSUE THREAD

**Issue by osa1** [gtk-rs/cairo#232]
Currently I can get a `&Surface` from a `&ImageSurface`, but I can't get a `&mut Surface` from `&mut ImageSurface` because there isn't an impl `AsMut<Surface> for ImageSurface`. This makes some operations impossible, e.g. I can't call `set_device_offset()` (added with #230) on an `ImageSurface`.

**Comment by sdroege** [comment]
Oops. Those should all be `&self` instead of `&mut self`, again interior mutability.

**Comment by osa1** [comment]
Sigh. Patch coming.

In general can we say that all `cairo` functions should be taking `&self` and never `&mut self` ?

**Comment by osa1** [comment]
@sdroege what about other `&mut self` `Surface`  methods like `set_device_scale` and `set_fallback_resolution` ?

**Comment by sdroege** [comment]
Same. The only exception in cairo is (IIRC) the one `ImageSurface` method that currently takes a mutable reference and gives access to the underlying data.

**Comment by osa1** [comment]
Interesting.. I don't see any mentions to thread safety in cairo documentation. How do you decide which ones should be `&mut self` and which ones should be `&self` ?

**Comment by sdroege** [comment]
> Interesting.. I don't see any mentions to thread safety in cairo documentation. How do you decide which ones should be `&mut self` and which ones should be `&self` ?

Apart from the image surface, none of the surfaces (and not the base class) are implementing `Send` so you can only ever use them on the original thread. That means that there are no problems with regard to threading, and the cairo API generally behaves like Rust interior mutability.

For the image surface you can only use that `&mut self` method if the refcount is 1 (i.e. no other thread can possibly have a reference to the surface), and by having the reference mutable there's also no way that you can pass the surface to another thread while also accessing the underlying data.

Now for the setters on the image surface... those seem indeed unsafe the way it is. You could call all that simultaneously on different threads and things will fail.

And the same is true for doing any actual drawing operations on the image surface.

----

So in summary, the image surface also can't be `Send`. None of the surfaces can unless we ensure that any operations that mutate data first check that the refcount is 1, but that will break other code (e.g. if we get a surface from GTK for drawing on it).

We should remove the `Send` impls from all surfaces until someone figures out a way how to implement this safely.

@GuillaumeGomez @EPashkin Am I missing something?

**Comment by GuillaumeGomez** [comment]
No, or at least I don't where you could be wrong.

**Comment by sdroege** [comment]
@osa1 I guess then add another commit to your PR to remove the `Send` impl from image (and PDF!) surface

**Comment by osa1** [comment]
What about methods that want `Send` arguments, like `set_mime_data` and `create_for_data`? I don't understand why they require `Send`.

**Comment by sdroege** [comment]
> What about methods that want `Send` arguments, like `set_mime_data` and `create_for_data`? I don't understand why they require `Send`.

They are not required anymore then, but you'll also have a hard time finding a type that fits in there and is not `Send` :)

**Comment by osa1** [comment]
Updated the PR.
